geochem_inverse_optimize.py

```diff
@@ -267,20 +319,145 @@ def get_downstream_prediction_dictionary(self) -> ElementData:
+        area_labels (np.array) : 2D array which matches upstream areas to labels
+        sample_network (nx.Digraph) : Network of sample_sites along drainage, with associated data
+
+    Attributes:
+        nx (int) : Number of columns in the grid
+        ny (int) : Number of rows in the grid
+        area_labels (np.array): 2D array which matches pixels to sample labels
+        node_arr (np.array): List of lists (dims: (ny,xs)) containing all InverseNodes
+        sites_to_nodes {str : list of InverseNode}: Dict mapping sample numbers to list of nodes in its upstream area
+    """
+
+    def __init__(self, nx: int, ny: int, area_labels: np.array, sample_network: nx.DiGraph) -> None:
+        self.area_labels = area_labels
+        if nx <= 0 or ny <= 0:
+            raise Exception("Warning: nx or ny cannot be negative")
+        xmax = area_labels.shape[1]
+        ymax = area_labels.shape[0]
+        if ny > ymax or nx > xmax:
+            raise Exception(
+                "Warning: desired resolution greater than that of DEM. \n Decrease resolution to resolve"
+            )
+        self.nx = nx
+        self.ny = ny
+        xstep = xmax / nx
+        ystep = ymax / ny
+        # The x and y coordinates on the DEM of the *centres* of the rectangular nodes
+        xs = np.linspace(start=xstep / 2, stop=xmax - xstep / 2, num=nx)
+        ys = np.linspace(start=ystep / 2, stop=ymax - ystep / 2, num=ny)
+        self.sites_to_nodes = defaultdict(list)
+        # Map area labels to sample numbers
+        area_label_to_sample_num = {
+            data["data"].label: node for node, data in sample_network.nodes(data=True)
+        }
+        self.node_arr = np.empty((ny, nx), dtype=object)
+        # Loop through a (nx, ny) grid
+        for i, x_coord in enumerate(xs):
+            for j, y_coord in enumerate(ys):
+                # Create an inversion node
+                node = InverseNode()
+                self.node_arr[j, i] = node
+                # Catch exception of left most nodes (used for roughness calculation)
+                if i != 0:
+                    # Point towards neighbour
+                    node.left_neighbour = self.node_arr[j, i - 1]
+                # Catch exception of upper most nodes
+                if j != 0:
+                    # Point towards neighbour (used for roughness calculation)
+                    node.top_neighbour = self.node_arr[j - 1, i]
+                # Set the sample number based off area map
+                # Sample number corresponds to sample downstream of the *centre* of the node
+                label = self.area_labels[int(y_coord), int(x_coord)]  # area label
+                # Only assign variables for nodes within the sampled area
+                if label == 0:
+                    node.sample_num = "NaN"
+                else:
+                    node.sample_num = area_label_to_sample_num[label]
+                    node.concentration = cp.Variable(pos=True)
```

Comment:
It might be better to set this up inside of `SampleNetwork` to more cleanly separate the optimization problem from arranging the data. Your call.

Alternatively, note that you could do a default initialization to a `cp.Variable` within the dataclass itself (probably using a default factory) and ever having a situation where the variable is not defined.

Variables that are not used in the optimization problem are ignored.

Reply:
Agree that it would be ideal to define this within the dataclass.

I tried to set up a default factory for `concentration` but requires a zero argument callable, which doesn't apply for our case as we're using the `pos=True` argument. Any idea how we could set that up?

Reply:
Probably use a lambda function.